pkg/gadgets/trace/dns/tracer/bpf/dns.c

```diff
@@ -213,52 +222,64 @@ int ig_trace_dns(struct __sk_buff *skb)
 	// spending useless cycles.
-	__builtin_memset(&event, 0, sizeof(event));
+	bpf_map_update_elem(&tmp_events, &zero, &empty_event, BPF_NOEXIST);
+	event = bpf_map_lookup_elem(&tmp_events, &zero);
+	if (!event)
+		return 0;
+	bpf_map_delete_elem(&tmp_events, &zero);
 
-	event.netns = skb->cb[0]; // cb[0] initialized by dispatcher.bpf.c
-	event.timestamp = bpf_ktime_get_boot_ns();
-	event.proto = proto;
-	event.dns_off = dns_off;
-	event.pkt_type = skb->pkt_type;
-	event.sport = sport;
-	event.dport = dport;
+	event->netns = skb->cb[0]; // cb[0] initialized by dispatcher.bpf.c
+	event->timestamp = bpf_ktime_get_boot_ns();
+	event->proto = proto;
+	event->dns_off = dns_off;
+	event->pkt_type = skb->pkt_type;
+	event->sport = sport;
+	event->dport = dport;
 
 	// The packet is DNS: Do a second pass to extract all the information we need
 	switch (h_proto) {
 	case ETH_P_IP:
-		event.af = AF_INET;
-		event.daddr_v4 = load_word(
+		event->af = AF_INET;
+		event->daddr_v4 = load_word(
 			skb, ETH_HLEN + offsetof(struct iphdr, daddr));
-		event.saddr_v4 = load_word(
+		event->saddr_v4 = load_word(
 			skb, ETH_HLEN + offsetof(struct iphdr, saddr));
 		// load_word converts from network to host endianness. Convert back to
 		// network endianness because inet_ntop() requires it.
-		event.daddr_v4 = bpf_htonl(event.daddr_v4);
-		event.saddr_v4 = bpf_htonl(event.saddr_v4);
+		event->daddr_v4 = bpf_htonl(event->daddr_v4);
+		event->saddr_v4 = bpf_htonl(event->saddr_v4);
 		break;
 	case ETH_P_IPV6:
-		event.af = AF_INET6;
+		event->af = AF_INET6;
 		if (bpf_skb_load_bytes(
 			    skb, ETH_HLEN + offsetof(struct ipv6hdr, saddr),
-			    &event.saddr_v6, sizeof(event.saddr_v6)))
+			    &event->saddr_v6, sizeof(event->saddr_v6)))
 			return 0;
 		if (bpf_skb_load_bytes(
 			    skb, ETH_HLEN + offsetof(struct ipv6hdr, daddr),
-			    &event.daddr_v6, sizeof(event.daddr_v6)))
+			    &event->daddr_v6, sizeof(event->daddr_v6)))
 			return 0;
 		break;
 	}
 
 	// Enrich event with process metadata
 	struct sockets_value *skb_val = gadget_socket_lookup(skb);
 	if (skb_val != NULL) {
-		event.mount_ns_id = skb_val->mntns;
-		event.pid = skb_val->pid_tgid >> 32;
-		event.tid = (__u32)skb_val->pid_tgid;
-		__builtin_memcpy(&event.task, skb_val->task,
-				 sizeof(event.task));
-		event.uid = (__u32)skb_val->uid_gid;
-		event.gid = (__u32)(skb_val->uid_gid >> 32);
+		event->mount_ns_id = skb_val->mntns;
+		event->pid = skb_val->pid_tgid >> 32;
+		event->tid = (__u32)skb_val->pid_tgid;
+		event->ppid = skb_val->ppid;
+		__builtin_memcpy(&event->comm, skb_val->task,
+				 sizeof(event->comm));
+		__builtin_memcpy(&event->pcomm, skb_val->ptask,
+				 sizeof(event->pcomm));
+		event->uid = (__u32)skb_val->uid_gid;
+		event->gid = (__u32)(skb_val->uid_gid >> 32);
+		bpf_probe_read_kernel_str(&event->cwd, sizeof(event->cwd),
+					  skb_val->cwd);
+		bpf_probe_read_kernel_str(&event->exepath,
+					  sizeof(event->exepath),
+					  skb_val->exepath);
```

Comment:
Any difference between this and `__builtin_memcpy` used above?

Reply:
`bpf_probe_read_kernel_str` copies until the NULL character and `__builtin_memcpy` copies a fixed buffer size without checking for NULL.

Also, with `__builtin_memcpy`, the compiler will automatically choose a faster implementation: unrolling the loop or not. With 16 bytes, it works fine, and I guess it will be faster than `bpf_probe_read_kernel_str` because we avoid the function call and we don't need to check for NULL. But using `__builtin_memcpy` for  4096 bytes will not work because the compiler will not unroll the loop and the verifier will reject the program.

A process name is set with [__set_task_comm](https://github.com/torvalds/linux/blob/88fac17500f4ea49c7bac136cf1b27e7b9980075/fs/exec.c#L1286) -> [strscpy_pad](https://github.com/torvalds/linux/blob/88fac17500f4ea49c7bac136cf1b27e7b9980075/include/linux/string.h#L123-L144) so it is guaranteed to be NULL-terminated and padded with zero until the end of the 16-byte buffer. So for process names, it is not necessary to use the slower `bpf_probe_read_kernel_str`.

Reply:
Thanks for the detailed explanation!